Add information on default interface members - added in C# 8.0

BillWagner opened the pull request:
Fixes #12490 
Fixes #16061 

Interfaces can now contain implementations of some member types.

- Include information on default interface members and what new member types are allowed in interfaces.
- Rewrite the opening to match the new uses of interfaces.
- Refer to the correct section of the language standard.
- Add a reference to the C# 8.0 default interface members specification.

BillWagner committed:
update warnings for DIM

For this commit, I checked all language reference articles that referred to an "interface". Those were all checked for any necessary updates.

The main "interface" article needs more work, and will be in the next commit.

BillWagner committed:
add information on default interface members

Interfaces can now contain implementations of some member types.

- Include information on default interface members and what new member types are allowed in interfaces.
- Rewrite the opening to match the new uses of interfaces.
- Refer to the correct section of the language standard.
- Add a reference to the C# 8.0 default interface members specification.

BillWagner committed:
fix build warning

Youssef1313 commented on docs/csharp/language-reference/keywords/interface.md:
```diff
@@ -1,6 +1,6 @@
 ---
 title: "interface - C# Reference"
-ms.date: 07/20/2015
+ms.date: 01/17/2019
```
```suggestion
ms.date: 01/17/2020
```

BillWagner replied:
Doh!

Youssef1313 commented on docs/csharp/language-reference/keywords/interface.md:
```diff
@@ -9,31 +9,40 @@ ms.assetid: 7da38e81-4f99-4bc5-b07d-c986b687eeba
 ---
 # interface (C# Reference)
```
Should "interface" be added to no-loc in the metadata?

pkulikov replied:
It looks that H1 (or C# keywords) are treated in some special way, for example:
https://docs.microsoft.com/ru-ru/dotnet/csharp/language-reference/keywords/interface

Also note that the C# keywords are not translated in the TOC.
So, I guess some "magic" is applied to the C# keywords met in the titles (and, thus, in TOCs)

Youssef1313 replied:
Yes it seems that it doesn't get localized in the heading. But in the text under the H1 it gets localized, which I think not desirable.

pkulikov replied:
If it's used as a keyword, it is usually code-fenced and not localized. However, when it's used as a term, it should be localized and not code-fenced.

pkulikov replied:
So, maybe, in this particular case, the title should be localized as well.

BillWagner replied:
This shouldn't be localized. I'll add it to the metadata.

BillWagner commented:
set the status back to work in progress. I think with a couple small edits, I can address three more issues related to language reference and DIM in this PR.

gewarren commented:
I'll remove my request until you're ready for review, if that's okay.

BillWagner committed:
Update static description

Interfaces may now declare static methods.

Fixes #16061

BillWagner committed:
Acrolinx recommendations

BillWagner commented:
@gewarren I've finished the remaining updates. This should be ready.

BillWagner committed:
Clarify property declaration

The same syntax for auto implemented properties declares a property without an implementation in an interface.

See https://github.com/dotnet/samples/pull/1917#r369281483

gewarren commented on docs/csharp/language-reference/keywords/static.md:
```diff
@@ -10,48 +10,46 @@ ms.assetid: 5509e215-2183-4da3-bab4-6b7e607a4fdf
 > The `static` keyword has more limited uses than in C++. To compare with the C++ keyword, see [Storage classes (C++)](/cpp/cpp/storage-classes-cpp#static).
 
-To demonstrate static members, consider a class that represents a company employee. Assume that the class contains a method to count employees and a field to store the number of employees. Both the method and the field do not belong to any instance employee. Instead they belong to the company class. Therefore, they should be declared as static members of the class.
+To demonstrate static members, consider a class that represents a company employee. Assume that the class contains a method to count employees and a field to store the number of employees. Both the method and the field don't belong to any instance employee. Instead they belong to the company class. They should be declared as `static` members of the class.
```
```suggestion
To demonstrate static members, consider a class that represents a company employee. Assume that the class contains a method to count employees and a field to store the number of employees. Both the method and the field don't belong to any one employee instance. Instead, they belong to the class of employees as a whole. They should be declared as `static` members of the class.
```

gewarren commented on docs/csharp/language-reference/keywords/interface.md:
```diff
@@ -1,39 +1,55 @@
+- Member declarations using the explicit interface implementation syntax.
+- Explicit access modifiers (the default access is [`public`](access-modifiers.md)).
+
+Interfaces may not contain instance state. While static fields are now permitted, instance fields are not permitted in interfaces. [Instance auto-properties](../../programming-guide/classes-and-structs/auto-implemented-properties.md) are not supported in interfaces, as they would implicitly declare a hidden field. This rule has a subtle effect on property declarations. In an interface declaration ...
```
```suggestion
Interfaces may not contain instance state. While static fields are now permitted, instance fields are not permitted in interfaces. [Instance auto-properties](../../programming-guide/classes-and-structs/auto-implemented-properties.md) are not supported in interfaces, as they would implicitly declare a hidden field. This rule has a subtle effect on property declarations. In an interface declaration, the following code does not declare an auto-implemented property as it does in a `class` or `struct`. Instead, it declares a property that doesn't have a default implementation but must be implemented in any type that implements the interface:
```

gewarren reviewed (approved):
Looks good overall! Thanks for adding the clarification. I was confused :)

BillWagner commented on docs/csharp/language-reference/keywords/interface.md:
```diff
@@ -1,39 +1,48 @@
 - [Events](event.md)
 
-An interface can inherit from one or more base interfaces.
+These preceding member declarations typically do not contain a body. Beginning with C# 8.0, an interface may declare a body, called a *default implementation*. Members with bodies permit the interface to provide a "default" implementation for the method in classes and structs that do not provide an overriding implementation. In addition, beginning with C# 8.0, an interface may include:
```
I'd like to get your thoughts here.  Providing a body for events, indexers and properties is legal, but in practice it's not generally useful. Declaring an instance property almost always implies declaring an instance field, which isn't legal. 

I can mention it, but any example would be contrived.

gewarren replied:
I think it's fine to leave it as-is then.

BillWagner commented on docs/csharp/language-reference/keywords/static.md:
````diff
@@ -10,48 +10,46 @@ ms.assetid: 5509e215-2183-4da3-bab4-6b7e607a4fdf
 ```
 
-While an instance of a class contains a separate copy of all instance fields of the class, there is only one copy of each static field.
+While an instance of a class contains a separate copy of all instance fields of the class, there's only one copy of each static field.
````
My intent was that all instances would be code-fenced. I'll check again.

BillWagner committed:
Apply suggestions from code review

Co-Authored-By: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>
Co-Authored-By: Genevieve Warren <gewarren@microsoft.com>

BillWagner committed:
code fence static

Except in title links, and the opening sentence, where the intent is clearly not as the keyword, but as the concept of static.
